docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+```cs
+int ComputeLastIndex<T>(nint elementOffset, Vector128<T> equals) where T : struct
+{
+    uint mostSignificantBits = equals.ExtractMostSignificantBits();
+
+    int index = 31 - BitOperations.LeadingZeroCount(mostSignificantBits); // 31 = 32 (bits in UInt32) - 1 (indexing from zero)
+
+    return (int)elementOffset + index;
+}
+```
+
+If we were using the `Load` overload that takes only the managed reference, we could use [Unsafe.ByteOffset<T>(ref T, ref T)](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafe.byteoffset) to calculate the element offset.
+
+```cs
+unsafe int ComputeFirstIndex<T>(ref T searchSpace, ref T current, Vector128<T> equals) where T : struct
+{
+    int elementOffset = (int)Unsafe.ByteOffset(ref searchSpace, ref current) / sizeof(T);
+
+    uint mostSignificantBits = equals.ExtractMostSignificantBits();
+    int index = BitOperations.TrailingZeroCount(mostSignificantBits);
+    
+    return elementOffset + index;
+}
+```
+
+### Comparison
+
+Beside equality checks, vector APIs allow for comparison. The `bool`-returning overloads return `true` when the given condition is true:
+
+```cs
+public static bool GreaterThanAll<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool GreaterThanAny<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool GreaterThanOrEqualAll<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool GreaterThanOrEqualAny<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool LessThanAll<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool LessThanAny<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool LessThanOrEqualAll<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static bool LessThanOrEqualAny<T>(Vector128<T> left, Vector128<T> right) where T : struct
+```
+
+Similarly to `Equals`, vector-returning overloads return a vector whose elements are all-bits-set or zero, depending on whether the corresponding elements in `left` and `right` meet the given condition.
+
+```cs
+public static Vector128<T> GreaterThan<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static Vector128<T> GreaterThanOrEqual<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static Vector128<T> LessThan<T>(Vector128<T> left, Vector128<T> right) where T : struct
+public static Vector128<T> LessThanOrEqual<T>(Vector128<T> left, Vector128<T> right) where T : struct
+```
+
+`ConditionalSelect` Conditionally selects a value from two vectors on a bitwise basis.
+
+```cs
+public static Vector128<T> ConditionalSelect<T>(Vector128<T> condition, Vector128<T> left, Vector128<T> right)
```

Comment:
You might show that this is basically `(left & condition) | (right & ~condition)`